### ISSUE TITLE
Implement basic authentication mechanism

### ISSUE THREAD

**Issue by docelic** [boomerdigital/solidus_graphql_api#14]
We should implement a basic authentication mechanism which operates through mutations and returns a JWT token to the user if login is successful.

The name of this mutation should be user_sign_in, so that we keep all user-related queries and mutations beginning with a common prefix of "user_*". This will then be easier to remember and query both via command line tools and GraphiQL interface, and will help us avoid method naming mess when we add more various queries and mutations.

**Comment by docelic** [comment]
A basically-working/prototype implementation of JWT authentication is in https://github.com/crystallabs/solidus_graphql_api/tree/jwt-authentication-1.

The primary things to fix are to read token: value from request rather than (currently) decoding a hard-coded token string, and to add more JWT fields (primarily token expiry, etc.) into the JWT token payload.

The same branch also implements a `me` query as a direct test of the functionality, and it looks like this:

Query:
```
me {
  id
  email
}
```

Response:
```
{
  "data": {
    "me": {
      "id": "1",
      "email": "admin@example.com"
    }
  }
}
```

**Comment by docelic** [comment]
Re. JWT in Solidus we should keep an eye on any progress in https://github.com/solidusio/solidus/pull/2808

**Comment by dhonig** [comment]
Yes. Ive been trying to coordinate this so perhaps we can just depend on
that gem for the JWT integration.

**Comment by mdesantis** [comment]
I'm acquiring informations about securing JWT authentications, it's a very interesting field (to me at least). This is the best resource I've found: https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/JSON_Web_Token_Cheat_Sheet_for_Java.md

Summarizing, this is what I believe we could implement:
  - token sidejacking: add user fingerprint to the token; it requires a combination of HttpOnly cookie and random string stored within the token (https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/JSON_Web_Token_Cheat_Sheet_for_Java.md#token-sidejacking)
  - token explicit revocation by the user: add a revoked tokens table (https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/JSON_Web_Token_Cheat_Sheet_for_Java.md#token-explicit-revocation-by-the-user)
  - token storage on client side: use sessionStorage container (https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/JSON_Web_Token_Cheat_Sheet_for_Java.md#token-storage-on-client-side)
  - preventing XSS attacks: add Content Security Policy headers (https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets_excluded/Content_Security_Policy_Cheat_Sheet.md)

**Comment by mdesantis** [comment]
I made some research, and it seems that `viewer` is somehow a de facto standard as GraphQL field name in order the return the current user. These GraphQL APIs use `viewer`:

- https://developer.github.com/v4/query/
- https://api.digitransit.fi/graphiql/finland
- https://hivdb.stanford.edu/page/graphiql/
- https://developer.mattermark.com/docs/graphql-schema
- https://www.universe.com/graphiql

**Comment by docelic** [comment]
Yeah, this concept with `viewer` is interesting because many people agree that almost everything in an API depends on who the "viewer" is.

In APIs which are completely based around "viewer", I think they use a model where GraphQL methods don't start from `QueryRoot` but from `viewer` (which can be a User model). In that case, `viewer` is resolved dynamically after the request comes in, when the application determines who the user/viewer is.
